components/server/src/main/java/com/hotels/styx/server/netty/codec/NettyToStyxRequestDecoder.java

```diff
@@ -61,80 +62,108 @@
+                if (content.isReadable()) {
+                    getContentProducer(ctx).newChunk(retain(content));
+                }
+                if (msg instanceof LastHttpContent) {
+                    getContentProducer(ctx).lastHttpContent();
                 }
+
             }
         } catch (BadRequestException ex) {
             throw ex;
         } catch (Exception ex) {
-            throw new BadRequestException(ex.getMessage() + " in " + httpObject, ex);
+            throw new BadRequestException(ex.getMessage() + " in " + msg, ex);
         }
     }
 
-    @Override
-    public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
-        if (this.producer != null) {
-            this.producer.notifySubscriber();
-        }
-        super.channelReadComplete(ctx);
-    }
-
     @Override
     public void channelInactive(ChannelHandlerContext ctx) throws Exception {
-        cleanUp();
+        TransportException cause = new TransportException(
+                "Connection to client lost: " + ctx.channel().remoteAddress());
+        getContentProducer(ctx).channelInactive(cause);
         super.channelInactive(ctx);
     }
 
     @Override
     public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
-        cleanUp();
-        if (cause instanceof TooLongFrameException) {
-            throw new BadRequestException(cause.getMessage(), cause);
-        }
+        getContentProducer(ctx).channelException(toStyxException(cause));
         super.exceptionCaught(ctx, cause);
     }
 
-    private void cleanUp() {
-        if (producer != null) {
-            producer.cleanUp();
+    private FlowControllingHttpContentProducer getNewContentProducer(ChannelHandlerContext ctx) {
+        this.producer = Optional.of(createProducer(ctx));
+        return this.producer.get();
+    }
+
+    private FlowControllingHttpContentProducer getContentProducer(ChannelHandlerContext ctx) {
+        if (!this.producer.isPresent()) {
+            this.producer = Optional.of(createProducer(ctx));
+        }
+        return this.producer.get();
+    }
+
+    private FlowControllingHttpContentProducer createProducer(ChannelHandlerContext ctx) {
+        String loggingPrefix = format("%s -> %s", ctx.channel().remoteAddress(), ctx.channel().localAddress());
+
+        return new FlowControllingHttpContentProducer(
+                () -> ctx.channel().read(),
+                () -> ctx.channel().config().setAutoRead(true),
+                cause -> { },
+                () -> scheduleResourcesTearDown(ctx),
+                format("%s, %s", loggingPrefix, ""),
+                null);
+    }
+
+    private void scheduleResourcesTearDown(ChannelHandlerContext ctx) {
+        ctx.channel().eventLoop().schedule(
+                () -> producer.ifPresent(FlowControllingHttpContentProducer::tearDownResources),
+                TEARDOWN_DELAY,
+                MILLISECONDS);
```

Comment:
Now looking back retrospectively, it seems bit funny that we tear down resources by generating a `ResponseTimeoutException` with "channelClosed" reason phrase. Let's think about improving this in the near future.

Reply:
By the same token, I now realise that the `scheduleTeardown` trigger should be different for uplink traffic.

The FSM was developed for downlink traffic. In downlink direction we cannot have two back-to-back responses without subscriber fully consuming the response body. Therefore in downlink we can use channel idleness to drive the `scheduleTeardown` event.

  * The FSM must go to the COMPLETED state before the channel can be reused for the next request. Because origin cannot respond anything before we send the another request, the cannel (TCP connection) would remain idle if the content subscriber never consumes the content. 

But for uplink, I believe, this assumption is not valid. It is possible for an origin to respond without consuming the request body at all. This response would allows a remote a sender to send a new request even if Styx never consumed the original request body. Therefore we cannot rely on channel idleness to clean up resources.

We need to think of something else that is guaranteed to trigger a clean up.